includes/payment-methods/class-upe-payment-gateway.php

```diff
@@ -867,4 +870,65 @@ private function is_enabled_for_saved_payments( $payment_method_id ) {
 		return $this->payment_methods[ $payment_method_id ]->is_reusable()
 			&& ( is_admin() || $this->payment_methods[ $payment_method_id ]->is_currency_valid() );
 	}
+
+	/**
+	 * Log UPE Payment Errors on Checkout.
+	 *
+	 * @throws Exception If nonce is not present or invalid or charge ID is empty or order not found.
+	 */
+	public function log_payment_error_ajax() {
+		try {
+			$is_nonce_valid = check_ajax_referer( 'wcpay_log_payment_error_nonce', false, false );
+			if ( ! $is_nonce_valid ) {
+				throw new Exception( 'Invalid request.' );
+			}
+
+			$charge_id = isset( $_POST['charge_id'] ) ? wc_clean( wp_unslash( $_POST['charge_id'] ) ) : '';
+			if ( empty( $charge_id ) ) {
+				throw new Exception( 'Charge ID cannot be empty.' );
+			}
+
+			// Get charge data from WCPay Server.
+			$get_charge_data = $this->payments_api_client->get_charge( $charge_id );
+			$order_id        = $get_charge_data['metadata']['order_id'];
+
+			// Validate Order ID and proceed with logging errors and updating order status.
+			$order = wc_get_order( $order_id );
+			if ( ! $order ) {
+				throw new Exception( 'Order not found. Unable to log error.' );
+			}
+
+			$order_note = sprintf(
+				WC_Payments_Utils::esc_interpolated_html(
+				/* translators: %1: the failed payment amount, %2: error message  */
+					__(
+						'A payment of %1$s <strong>failed</strong> to complete with the following message: <code>%2$s</code>.',
+						'woocommerce-payments'
+					),
+					[
+						'strong' => '<strong>',
+						'code'   => '<code>',
+					]
+				),
+				WC_Payments_Explicit_Price_Formatter::get_explicit_price( wc_price( $order->get_total(), [ 'currency' => $order->get_currency() ] ), $order ),
```

Comment:
Can you store this is variable above and use it here.

Reply:
Sure. Stored both the amount and error message as new variables for readability.